src/Shared/JSInterop/JSCallResultTypeHelper.cs

```diff
@@ -6,7 +6,9 @@ namespace Microsoft.JSInterop
     internal static class JSCallResultTypeHelper
     {
         public static JSCallResultType FromGeneric<TResult>()
-            => typeof(TResult) == typeof(IJSObjectReference) || typeof(TResult) == typeof(IJSInProcessObjectReference) ?
+            => typeof(TResult) == typeof(IJSObjectReference)
+            || typeof(TResult) == typeof(IJSInProcessObjectReference)
+            || typeof(TResult) == typeof(IJSUnmarshalledObjectReference) ?
```

Comment:
Here's a minor suggestion - I'll leave it to you whether to do it or not. It might not make any difference.

In the most common case, all three of these checks will give `false`. Is it worth optimizing for that common case by reducing it to one check in that case? For example:

```cs
(typeof(TResult).Assembly == thisAssembly) && (... existing checks ...)
````

I guess we could even 100% block any confusion about using the concrete implementation types (`InvokeAsync<JSObjectReference>`) like:

```cs
if (typeof(TResult).Assembly == thisAssembly)
{
    if (it is any of the three accepted interface types)
    {
        return JSCallResultType.JSObjectReference;
    }
    else
    {
        throw InvalidArgumentException("JS interop cannot supply an instance of type {TResult}");
    }
}
else
{
    return JSCallResultType.Default;
}

```

Reply:
@SteveSandersonMS and I discussed this offline - I am going to file a separate issue for this since it's not crucial it gets implemented right now.